cpp/tests/gis/point_in_polygon_test.cu

```diff
@@ -0,0 +1,181 @@
+    T orientation(T p1_x, T p1_y, T p2_x, T p2_y, T p3_x, T p3_y)
+    {
+	    return ((p2_y - p1_y) * (p3_x - p2_x) - (p2_x - p1_x) * (p3_y - p2_y));
+    }
+
+    // pip host function
+    std::vector<int8_t> compute_reference_pip()
+    {   
+        size_t total_points = point_lats.size();
+        std::vector<int8_t> h_inside_polygon(total_points, -1);
+
+        for (size_t id_point = 0; id_point < total_points; ++id_point)
+        {
+            T point_lat = point_lats[id_point];
+            T point_lon = point_lons[id_point];
+            int count = 0;
+
+            for (size_t poly_idx = 0; poly_idx < polygon_lats.size() - 1; ++poly_idx)
+            {
+                if(polygon_lons[poly_idx] <= point_lon && point_lon < polygon_lons[poly_idx + 1])
+                {
+                    if (orientation(polygon_lons[poly_idx], polygon_lats[poly_idx], polygon_lons[poly_idx + 1], polygon_lats[poly_idx + 1], point_lon, point_lat) > 0)
+                    {
+                        count++;
+                    }
+                }
+                else if (point_lon <= polygon_lons[poly_idx] && polygon_lons[poly_idx + 1] < point_lon) 
+                {
+                    if (orientation(polygon_lons[poly_idx], polygon_lats[poly_idx], polygon_lons[poly_idx + 1], polygon_lats[poly_idx + 1], point_lon, point_lat) > 0)
+                    {
+                        count++;
+                    }
+                }
+            }
+
+            if ((count > 0) && (count % 2 == 0)) h_inside_polygon[id_point] = 1;
+		    else h_inside_polygon[id_point] = 0;
+        }
+
+        return h_inside_polygon;
+    }
+
+    std::vector<int8_t> compute_gdf_pip()
+    {
+        cudf::test::column_wrapper<T> polygon_lat_wrapp{polygon_lats};
+        cudf::test::column_wrapper<T> polygon_lon_wrapp{polygon_lons};
+        cudf::test::column_wrapper<T> point_lat_wrapp{point_lats};
+        cudf::test::column_wrapper<T> point_lon_wrapp{point_lons};
+
+        gdf_column inside_poly_col = cudf::point_in_polygon( *(polygon_lat_wrapp.get()), *(polygon_lon_wrapp.get()), 
+                                                             *(point_lat_wrapp.get()), *(point_lon_wrapp.get()) );
+
+        std::vector<int8_t> host_inside_poly(point_lats.size());
+
+        EXPECT_EQ(cudaMemcpy(host_inside_poly.data(), inside_poly_col.data, inside_poly_col.size * sizeof(int8_t), cudaMemcpyDeviceToHost), cudaSuccess);
+    
+        return host_inside_poly;
+    }
+};
+
+// Geographical data are real numeric (int8_t is lesser than the min_max longitude value)
+typedef testing::Types<int16_t, int32_t, int64_t, float, double> NumericTypes;
+
+TYPED_TEST_CASE(PIPTest, NumericTypes);
+
+TYPED_TEST(PIPTest, InsidePolygonClockwise)
+{
+    // Latitudes polygon, longitudes polygon, latitudes of query points, longitudes of query points, print = false
+    this->create_input({-10, -10, 10, 10, -10}, {10, -10, -10, 10, 10}, {4, 5, 2, 6, 2, 8}, {2, 6, 5, 8, 1, 3}, false);
```

Comment:
As suspected, your tests are insufficient. ALL points in this test are inside the polygon. ALL points in the other test are outside the polygon. This might explain why your tests are able to pass even though your kernel only ever tests the first point.

Please add tests that mix up inside and outside points.  Also add tests that generate points randomly rather than manually.

These tests also only test a small number of points in low-vertex-count polygons. Please add one test that tests a lot of points and/or a big polygon.

You should also probably test degenerate polygons. What happens if the polygon self-intersects? What if it has colocated points? What if all vertices are collinear? It should be robust to all these cases, and the expected output should be documented.

Reply:
* I modified some parts of code about you feedback.
* I created a new small test mixing up both inside and outside points how did you recommend.
* I created a test generating 1000 random  points, as well.
* I created a test for a polygon more complex (similar to the second example in the documentation of the point_i_polygon_kernel (cpp/src/gis/point_in_polygon.cu) ) with random query points.
* I also created a test (just for my local) with a real polygon (extension of a city in a country) and the test was ok.

* Polygon can be built in clockwise and counter-clockwise, works well for both.
* What do you mean with colocated points? Points in the boundary, maybe?
* If all vertices are collinear, then it doesn't a polygon it is rather a multiline.
* At this moment the code doesn't work completely for degenerate polygons.  I mean polygon must not have holes neither intersect with itself.

In cpp/include/gis.hpp now there are more specifications about the first GIS algorithm.

Reply:
What I mean is that your function must act responsibly for degenerate polygons. For example, if all points are colocated at the same point, it is a polygon of zero area -- what is the expected result?  If all points are collinear it is a line -- what is the expected result? Also as you mentioned, polygons with holes. These polygons may not be valid, but you should at least document and test these cases to ensure the expected behavior is achieved.

Reply:
I created more test for those cases.  Thanks for the constants feedback. So about what I plan to do with the last case (for polygon intersecting with itself), I will document the shortcoming for now. I have been working in a second GIS algorithm (compute of the perimeter of spherical polygons) but I will be thinking for the next version of the function point-in-polygon (to ensure it will be more robust).